plugins/pulp_docker/plugins/importers/sync.py

```diff
@@ -187,122 +130,62 @@ def process_main(self):
-                                            model.relative_path)
-
+        super(SaveUnitsStep, self).__init__(step_type=constants.SYNC_STEP_SAVE)
+        self.description = _('Saving Manifests and Blobs')
 
-class SaveUnitsStep(PluginStep):
-    def __init__(self, working_dir):
+    def get_iterator(self):
         """
-        :param working_dir: full path to the directory into which blob files
-                            are downloaded. This directory should contain one
-                            directory for each docker blob, with the ID of the
-                            docker blob as its name.
-        :type  working_dir: basestring
+        Return an iterator that will traverse the list of Units that were downloaded.
+
+        :return: An iterable containing the Blobs and Manifests that were downloaded and are new to
+                 Pulp.
+        :rtype:  iterator
         """
-        super(SaveUnitsStep, self).__init__(
-            step_type=constants.SYNC_STEP_SAVE, plugin_type=constants.IMPORTER_TYPE_ID,
-            working_dir=working_dir)
-        self.description = _('Saving manifests and blobs')
+        return iter(self.parent.step_get_local_manifests.units_to_download +
+                    self.parent.step_get_local_blobs.units_to_download)
 
-    def process_main(self):
+    def process_main(self, item=None):
         """
-        Gets an iterable of units that were downloaded from the parent step,
+        Gets an iterable of Units that were downloaded from the parent step,
         moves their files into permanent storage, and then saves the unit into
         the database and into the repository.
-        """
-        _logger.debug(self.description)
-        # Save the Manifests
-        for unit_key in self.parent.step_get_metadata.step_get_local_units.units_to_download:
-            model = self.parent.step_get_metadata.manifests[unit_key['digest']]
-            unit = self.get_conduit().init_unit(model.TYPE_ID, model.unit_key, model.metadata,
-                                                model.relative_path)
-            self._move_file(unit)
-            _logger.debug('saving manifest %s' % model.digest)
-            self.get_conduit().save_unit(unit)
 
-        # Save the Blobs
-        for unit_key in self.parent.step_get_local_units.units_to_download:
-            model = models.Blob(unit_key['digest'])
-            unit = self.get_conduit().init_unit(model.TYPE_ID, model.unit_key, model.metadata,
-                                                model.relative_path)
-            self._move_file(unit)
-            _logger.debug('saving Blob %s' % unit_key)
-            self.get_conduit().save_unit(unit)
-
-    def _move_file(self, unit):
-        """
-        For the given unit, move its associated file from the working
-        directory to its permanent location.
+        This method will be called exactly once without an item argument if the iterator returned by
```

Comment:
This fix does address the coding error, but does not address the pattern that I referenced. The problem is that `process()`'s treatment of `_process_block()`, and thus the purpose of `_process_block()`, is dependent on other internal state of the object. This causes confusion as the presence or absence of some other attribute (the generator) causes `_process_block()` to be run through iteration, or to be run once with no argument. Additionally, the signature of `_process_block()` depends on this state.

The coding error did exacerbate the significance of the anti-pattern by requiring the if statement below, but the deeper problem is still present. The two meanings of `_process_block()` are an indication that there ought to be two separate methods - one to process each item from the generator and one to be run once during the step. We should change this method signature to `def _process_item(self, item)` (note that item is required), or a similar name to indicate that it is a different thing than `_process_block`.

All of this said, your commit did remove the "exacerbation" of the problem that is present here and so you are correct that the if statement is not needed here any longer. Future users of the step system will continue to stumble upon this problem unless we fix the core of the issue. I will also change the signature to require the item.

You are correct. It is well known that I dislike this system, and my comment was stepping beyond what belongs in the code. I will remove it.